src/chain/store/index.rs

```diff
@@ -151,139 +62,186 @@ impl<DB: Blockstore> ChainIndex<DB> {
-        to: ChainEpoch,
-    ) -> Result<Arc<Tipset>, Error> {
-        self.walk_back(from, to)
+    /// Iterate from the given tipset to genesis. Missing tipsets cut the chain short.
+    pub fn chain(&self, from: Arc<Tipset>) -> impl Iterator<Item = Arc<Tipset>> + '_ {
+        itertools::unfold(Some(from), move |tipset| {
+            tipset.take().map(|child| {
+                *tipset = self.load_tipset(child.parents()).ok();
+                child
+            })
+        })
     }
+}
 
-    /// Fills cache with look-back entry, and returns inserted entry.
-    fn fill_cache(&self, tsk: TipsetKeys) -> Result<Arc<LookbackEntry>, Error> {
-        let tipset = self.load_tipset(&tsk)?;
-
-        if tipset.epoch() == 0 {
-            return Ok(Arc::new(LookbackEntry {
-                tipset,
-                parent_height: 0,
-                target_height: Default::default(),
-                target: Default::default(),
-            }));
-        }
-
-        let parent = self.load_tipset(tipset.parents())?;
-        let r_height = self.round_height(tipset.epoch()) - SKIP_LENGTH;
-
-        let parent_epoch = parent.epoch();
-        let skip_target = if parent.epoch() < r_height {
-            parent
-        } else {
-            self.walk_back(parent, r_height)?
-        };
+#[cfg(test)]
+mod tests {
+    use std::sync::atomic::{AtomicU64, Ordering};
 
-        let lbe = Arc::new(LookbackEntry {
-            tipset,
-            parent_height: parent_epoch,
-            target_height: skip_target.epoch(),
-            target: skip_target.key().clone(),
-        });
+    use super::*;
+    use crate::blocks::BlockHeader;
+    use crate::db::MemoryDB;
+    use crate::utils::db::CborStoreExt;
 
-        self.skip_cache.lock().put(tsk, lbe.clone());
-        Ok(lbe)
+    fn persist_tipset(tipset: &Tipset, db: &impl Blockstore) {
+        for block in tipset.blocks() {
+            db.put_cbor_default(block).unwrap();
+        }
     }
 
-    /// Rounds height epoch to nearest sparse cache index epoch.
-    fn round_height(&self, height: ChainEpoch) -> ChainEpoch {
-        (height / SKIP_LENGTH) * SKIP_LENGTH
+    fn genesis_tipset() -> Tipset {
+        Tipset::from(BlockHeader::default())
     }
 
-    /// Gets the closest rounded sparse index and returns the loaded tipset at
-    /// that index.
-    fn round_down(&self, ts: Arc<Tipset>) -> Result<Arc<Tipset>, Error> {
-        let target = self.round_height(ts.epoch());
-
-        self.walk_back(ts, target)
+    fn tipset_child(parent: &Tipset, epoch: ChainEpoch) -> Tipset {
+        // Use a static counter to give all tipsets a unique timestamp
+        static COUNTER: AtomicU64 = AtomicU64::new(0);
+        let n = COUNTER.fetch_add(1, Ordering::Relaxed);
+        Tipset::from(
+            BlockHeader::builder()
+                .parents(parent.key().clone())
+                .epoch(epoch)
+                .timestamp(n)
+                .build()
+                .unwrap(),
```

Comment:
new issue: this builder shouldn't exist. Use something like tidy-builder instead

Reply:
#3229